src/ucs/stats/stats.c

```diff
@@ -112,6 +117,76 @@ static void ucs_stats_clean_node(ucs_stats_node_t *node) {
     ucs_list_del(&node->type_list);
 }
 
+static void ucs_stats_free_class(ucs_stats_class_t *cls)
+{
+    unsigned i;
+
+    for (i = 0; i < cls->num_counters; i++) {
+        ucs_free((void*)cls->counter_names[i]);
+    }
+
+    ucs_free((void*)cls->name);
+    ucs_free(cls);
+}
+
+static ucs_stats_class_t *ucs_stats_dup_class(ucs_stats_class_t *cls)
+{
+    ucs_stats_class_t *dup;
+
+    dup = ucs_calloc(1, sizeof(*cls) + sizeof(*cls->counter_names) * cls->num_counters,
+                     "ucs_stats_class_dup");
+    if (!dup) {
+        ucs_warn("failed to allocate statistics class");
+        goto err;
+    }
+
+    dup->name = ucs_strdup(cls->name, "ucs_stats_class_t name");
+    if (!dup->name) {
+        ucs_warn("failed to copy statistics class name");
+        goto err_free;
+    }
+
+    for (dup->num_counters = 0; dup->num_counters < cls->num_counters; dup->num_counters++) {
+        dup->counter_names[dup->num_counters] = ucs_strdup(cls->counter_names[dup->num_counters],
+                                                           "ucs_stats_class_t counter");
+        if (!dup->counter_names[dup->num_counters]) {
+            ucs_warn("failed to copy statistics counter name");
+            goto err_free;
+        }
+    }
+
+    return dup;
+
+err_free:
+    ucs_stats_free_class(dup);
+err:
+    return NULL;
+}
+
+static ucs_stats_class_t *ucs_stats_get_class(ucs_stats_class_t *cls)
+{
+    ucs_stats_class_t *dup;
+    khiter_t iter;
+    int r;
+
+    iter = kh_get(ucs_stats_cls, &ucs_stats_context.cls, cls->name);
+
+    if (iter != kh_end(&ucs_stats_context.cls)) {
+        return kh_val(&ucs_stats_context.cls, iter);
+    }
+
+    dup = ucs_stats_dup_class(cls);
+
+    if (dup == NULL) {
+        return NULL;
+    }
+
+    iter = kh_put(ucs_stats_cls, &ucs_stats_context.cls, cls->name, &r);
```

Comment:
must NOT pass cls->name because it can be released
need to put dup->name as hash key

Reply:
right, missed that strings are released too
fixed